contracts/cw-croncat/src/manager.rs

```diff
@@ -211,6 +177,105 @@ impl<'a> CwCroncat<'a> {
+            if let Some(gas_limit) = action.gas_limit {
+                sub_msgs.push(sub_msg.with_gas_limit(gas_limit));
+            } else {
+                sub_msgs.push(sub_msg);
+            }
+        }
+
+        // TODO: Add supported msgs if not a SubMessage?
+        // Add the messages, reply handler responsible for task rescheduling
+        let final_res = Response::new()
+            .add_attribute("method", "proxy_call")
+            .add_attribute("agent", info.sender)
+            .add_attribute("slot_id", slot_id.to_string())
+            .add_attribute("slot_kind", format!("{:?}", slot_type))
+            .add_attribute("task_hash", task.to_hash())
+            .add_submessages(sub_msgs);
+        Ok(final_res)
+    }
+
+    /// Executes a task based on the current task slot
+    /// Computes whether a task should continue further or not
+    /// Makes a cross-contract call with the task configuration
+    /// Called directly by a registered agent
+    pub fn proxy_call_with_rules(
+        &mut self,
+        deps: DepsMut,
+        info: MessageInfo,
+        env: Env,
+        task_hash: String,
+    ) -> Result<Response, ContractError> {
+        self.check_ready_for_proxy_call(deps.as_ref(), &info)?;
+
+        self.check_agent(deps.as_ref(), &info)?;
+
+        let some_task = self
+            .tasks_with_rules
+            .may_load(deps.storage, task_hash.to_owned().into_bytes())?;
+        if some_task.is_none() {
+            return Err(ContractError::NoTaskFound {});
+        }
+        let task = some_task.unwrap();
+
+        self.check_bank_msg(deps.as_ref(), &info, &env, &task)?;
+
+        // Check rules
+        let rules = task.rules.as_ref().expect("No rules");
+        for rule in rules {
+            let res: (bool, Option<String>) =
+                deps.querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+                    contract_addr: rule.contract_addr.clone().to_string(),
+                    msg: rule.msg.clone(),
+                }))?;
+            if !res.0 {
+                return Err(ContractError::RulesNotReady {});
+            }
+        }
+
+        // Check that this task can be executed in current slot
+        let task_ready = if let Ok(Some(block)) = self
+            .block_slots_rules
+            .may_load(deps.storage, task_hash.clone().into_bytes())
+        {
+            env.block.height >= block
+        } else if let Ok(Some(time)) = self
+            .time_slots_rules
+            .may_load(deps.storage, task_hash.clone().into_bytes())
+        {
+            env.block.time.nanos() >= time
+        } else {
+            // This shouldn't happen
+            return Err(ContractError::CustomError {
+                val: "Task doesn't have block or time slot".to_string(),
+            });
+        };
+        if !task_ready {
+            return Err(ContractError::CustomError {
+                val: "Wrong slot for this task".to_string(),
```

Comment:
Feels like we could have a different error message, unless I'm misunderstanding:

```suggestion
                val: "Task is not ready".to_string(),
```

Reply:
We have `Err(ContractError::RulesNotReady {})` if one of the rules failed. But here we were checking whether the task can be executed in the current clot, so I think the error should be `Wrong slot for this task` or something similar